pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -225,34 +243,138 @@ func newControllerNode(cfg VSphereConfig) (*VSphere, error) {
 			return nil, err
 		}
 	}
-	vSphereConn := vclib.VSphereConnection{
-		Username:          cfg.Global.User,
-		Password:          cfg.Global.Password,
-		Hostname:          cfg.Global.VCenterIP,
-		Insecure:          cfg.Global.InsecureFlag,
-		RoundTripperCount: cfg.Global.RoundTripperCount,
-		Port:              cfg.Global.VCenterPort,
+
+	vsphereInstanceMap := make(map[string]*VSphereInstance)
+	// Check if the vsphere.conf is in old format. In this
+	// format the cfg.VirtualCenter will be nil or empty.
+	if cfg.VirtualCenter == nil || len(cfg.VirtualCenter) == 0 {
+		glog.V(4).Infof("Config is not per virtual center.")
+		datacenters := ""
+		if cfg.Global.Datacenters != "" {
+			datacenters = cfg.Global.Datacenters
+		} else {
+			// cfg.Global.Datacenter is deprecated, so giving it the last preference.
+			datacenters = cfg.Global.Datacenter
+		}
+		if cfg.Global.User == "" {
+			glog.Error("Global.User is empty!")
+			return nil, errors.New("Global.User is empty!")
+		}
+		if cfg.Global.Password == "" {
+			glog.Error("Global.Password is empty!")
+			return nil, errors.New("Global.Password is empty!")
+		}
+
+		vcConfig := VirtualCenterConfig{
+			User:              cfg.Global.User,
+			Password:          cfg.Global.Password,
+			VCenterPort:       cfg.Global.VCenterPort,
+			Datacenters:       datacenters,
+			InsecureFlag:      cfg.Global.InsecureFlag,
+			RoundTripperCount: cfg.Global.RoundTripperCount,
+			WorkingDir:        cfg.Global.WorkingDir,
+		}
+
+		vSphereConn := vclib.VSphereConnection{
```

Comment:
Looks like VSphereConnection and VirtualCenterConfig are more or less duplicate.
Any strong reason to keep the duplication?

Reply:
VirtualCenterConfig represents the configuration as specified in vsphere.conf file. VSphereConnection is a domain object that maintains connection to VC. Also note that we have additional fields in VSphereConnection that are not in VirtualCenterConfig like Hostname and GoVmomiClient.